pkg/rbac/userData.go

```diff
@@ -18,41 +18,114 @@ import (
+	csResources []resource            // Cluster-scoped resources on hub the user has list access.
 	nsResources map[string][]resource // Namespaced resources on hub the user has list access.
 
-	// Internal fields to manage the cache.
-	// clustersErr       error      // Error while updating clusters data.
-	// clustersLock      sync.Mutex // Locks when clusters data is being updated.
-	// clustersUpdatedAt time.Time  // Time clusters was last updated.
-	err     error      // Error while getting user data from cache
-	csrErr  error      // Error while updating cluster-scoped resources data.
-	csrLock sync.Mutex // Locks when cluster-scoped resources data is being updated.
-	// csrUpdatedAt time.Time  // Time cluster-scoped resources was last updated.
+	err          error      // Error while getting user data from cache
+	csrErr       error      // Error while updating cluster-scoped resources data.
+	csrLock      sync.Mutex // Locks when cluster-scoped resources data is being updated.
+	csrUpdatedAt time.Time  // Time cluster-scoped resources was last updated.
 	nsrErr       error      // Error while updating namespaced resources data.
 	nsrLock      sync.Mutex // Locks when namespaced resources data is being updated.
 	nsrUpdatedAt time.Time  // Time namespaced resources was last updated.
-
-	// impersonate *kubernetes.Interface // client with impersonation config
+	authzClient  v1.AuthorizationV1Interface
 }
 
-func (cache *Cache) GetUserData(ctx context.Context, clientToken string) (*userData, error) {
-	var user userData
+func (cache *Cache) GetUserData(ctx context.Context, clientToken string, authzClient v1.AuthorizationV1Interface) (*userData, error) {
+	var user *userData
 	uid := cache.tokenReviews[clientToken].tokenReview.Status.User.UID //get uid from tokenreview
 	cache.usersLock.Lock()
 	defer cache.usersLock.Unlock()
 	cachedUserData, userDataExists := cache.users[uid] //check if userData cache for user already exists
-	if userDataExists {
+	// UserDataExists and its valid
+	if userDataExists && userCacheValid(cachedUserData) {
 		klog.V(5).Info("Using user data from cache.")
-		return cachedUserData, user.err
-
+		return cachedUserData, nil
+	} else {
+		// User not in cache , Initialize and assign to the UID
+		user = &userData{}
+		cache.users[uid] = user
+		// We want to setup the client if passed, this is only for unit tests
+		if authzClient != nil {
+			user.authzClient = authzClient
+		}
 	}
-
 	userData, err := user.getNamespacedResources(cache, ctx, clientToken)
+
+	// Get cluster scoped resources for the user
+	// TO DO : Make this parallel operation
+	if err == nil {
+		klog.V(5).Info("No errors on namespacedresources not present ....Computing now", user)
+		userData, err = user.getClusterScopedResources(cache, ctx, clientToken)
+	}
 	return userData, err
 
 }
 
+func userCacheValid(user *userData) bool {
+	if (time.Now().Before(user.csrUpdatedAt.Add(time.Duration(config.Cfg.UserCacheTTL) * time.Millisecond))) &&
```

Comment:
Can we have a comment here or update the comments in user cache clarifying that these `updatedAt` fields are the time when resources were authorized last.